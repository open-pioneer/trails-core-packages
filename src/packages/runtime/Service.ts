// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { PackageIntl } from "./i18n";

/** Lifecycle hooks supported by the service interface. */
export interface ServiceLifecycleHooks {
    /**
     * Destroys the component.
     * Should clean up all resources used by the component.
     */
    destroy?(): void;

    /*
     * Just used so that this interface is not empty.
     * TypeScript would otherwise throw an 'no common properties' error
     * if one tries to implement this interface without providing any
     * lifecycle hooks.
     */
    toString(): string;
}

export type Service<Interface extends {} = {}> = ServiceLifecycleHooks & Interface;

/**
 * Represents metadata for an injected reference.
 */
export interface ReferenceMeta {
    /**
     * The unique service id of the injected service.
     */
    serviceId: string;
}

/**
 * Contains metadata about injected references.
 * The key for the reference metadata is the reference's name.
 */
export type ReferencesMeta<References extends {}> = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [referenceName in keyof References]: References[referenceName] extends any[]
        ? ReferenceMeta[]
        : ReferenceMeta;
};

/**
 * Options passed to a service's constructor.
 */
export type ServiceOptions<References extends {} = {}> = {
    /**
     * References to other services.
     * These are injected by the runtime and match the declared references of the service.
     */
    references: References;

    /**
     * Metadata about the injected references.
     * The object contains one entry for every reference injected via `references` (under the same key).
     *
     * If a single service has been injected, `referencesMeta[key]` will be of type {@link ReferenceMeta}.
     * If multiple services have been injected, an array of {@link ReferenceMeta} will be provided instead,
     * where the indices match the order in the injected references array.
     */
    referencesMeta: ReferencesMeta<References>;

    /**
     * Configuration properties.
     *
     * Default values are taken from the package's configuration, but these can be overridden
     * by the application.
     */
    properties: Record<string, unknown>;

    /**
     * A reference to the shared i18n object of the current package.
     */
    intl: PackageIntl;
};

/**
 * A service constructor returns a service instance when calling `new`.
 */
export type ServiceConstructor<References extends {} = {}, Interface extends {} = {}> = {
    new (options: ServiceOptions<References>): Service<Interface>;
};
