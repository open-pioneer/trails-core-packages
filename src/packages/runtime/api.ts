// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { DeclaredService } from "./DeclaredService";

/* eslint-disable @typescript-eslint/no-explicit-any */
export type ApiMethod = (...args: any[]) => any;

/**
 * A record of exposed functions.
 */
export type ApiMethods = Record<string, ApiMethod>;

/**
 * Provides a set of methods to the application's web component API.
 */
export interface ApiExtension {
    /**
     * Returns a set of methods that will be added to the web component's API.
     */
    getApiMethods(): Promise<ApiMethods> | ApiMethods;
}

/**
 * A service provided by the system.
 * Used by the runtime to assemble the public facing API.
 */
export interface ApiService extends DeclaredService<"runtime.ApiService"> {
    /**
     * Called by the runtime to gather methods that should be available from the web component's API.
     */
    getApi(): Promise<ApiMethods>;
}

/**
 * A service provided by the system, useful for accessing values that are global to the application.
 */
export interface ApplicationContext extends DeclaredService<"runtime.ApplicationContext"> {
    /**
     * The web component's host element.
     * This dom node can be accessed by the host site.
     */
    getHostElement(): HTMLElement;

    /**
     * The current web component's shadow root.
     * This shadow root is located inside the host element.
     */
    getShadowRoot(): ShadowRoot;

    /**
     * The node containing the rest of the application _inside_ the current web component's shadow dom.
     */
    getApplicationContainer(): HTMLElement;

    /**
     * Returns the current locale of the application.
     *
     * E.g. `"de-DE"`
     */
    getLocale(): string;

    /**
     * Changes the application's locale.
     * `locale` must be one of the supported locales, see {@link getSupportedLocales()} or `undefined` (for automatic locale).
     * Note that `locale` does not need to be a precise match, e.g. `"de-DE"` is also valid if `"de"` is supported.
     *
     * > NOTE: This method will currently trigger a full restart of the application.
     * > Altering the locale on the fly is possible in theory but has not been implemented yet.
     */
    setLocale(locale: string | undefined): void;

    /**
     * Returns the locales supported by the application, i.e.
     * the locales that have associated i18n messages.
     *
     * For example: `["de", "en"]`
     */
    getSupportedLocales(): readonly string[];
}

/**
 * Implementing this interface allows a service to be notified on certain
 * application-wide lifecycle events.
 *
 * **Experimental**. This interface is not affected by semver guarantees.
 * It may change (or be removed) in a future minor release.
 */
export interface ApplicationLifecycleListener
    extends DeclaredService<"runtime.ApplicationLifecycleListener"> {
    /**
     * Called after all services required by the application have been started.
     */
    afterApplicationStart?(): void;

    /**
     * Called during the application shutdown just before services will be destroyed.
     */
    beforeApplicationStop?(): void;
}
