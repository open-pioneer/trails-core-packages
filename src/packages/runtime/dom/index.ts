// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
export type RootNode = Document | ShadowRoot;

/**
 * Class of the <div> that contains the entire application.
 */
export const APP_ROOT_CLASS = "pioneer-root";

export function isShadowRoot(node: RootNode): node is ShadowRoot {
    return node.nodeType === Node.DOCUMENT_FRAGMENT_NODE && "host" in node;
}

export function isDocument(node: RootNode): node is Document {
    return !isShadowRoot(node);
}

/**
 * Returns the parent node for <style> elements.
 */
export function getStylesRoot(node: RootNode): ShadowRoot | HTMLHeadElement {
    return isShadowRoot(node) ? node : node.head;
}

/**
 * Creates the node that contains all other application content.
 */
export function createAppRoot(locale: string): HTMLDivElement {
    // Setup application root node in the shadow dom
    const container = document.createElement("div");
    container.classList.add("pioneer-root");
    container.style.minHeight = "100%";
    container.style.height = "100%";
    if (locale) {
        container.lang = locale;
    }
    return container;
}

/**
 * Returns the styles for host element (the web component node).
 */
export function getBuiltinStyles(rootNode: RootNode, hostElement: HTMLElement | undefined) {
    // Prevent inheritance of certain css values and normalize to display: block by default.
    // See https://open-wc.org/guides/knowledge/styling/styles-piercing-shadow-dom/
    // NOTE: layer base comes from chakra (used for css resets etc).
    const selector =
        isShadowRoot(rootNode) || !hostElement ? ":host" : hostElement.tagName.toLowerCase();
    return `@layer base { ${selector} { all: initial; display: block; } }`;
}
