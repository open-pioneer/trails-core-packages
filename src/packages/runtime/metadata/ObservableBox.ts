// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { EventEmitter, Resource } from "@open-pioneer/core";

export interface ObservableBox<T> {
    /** The current value. */
    readonly value: T;

    /**
     * Changes the current value.
     *
     * Note that this function is only present during development to facilitate hot reloading.
     */
    setValue?(value: T): void;

    /**
     * Register a listener callback that will be invoked
     * when the value changes.
     *
     * Note that this function is only present during development to facilitate hot reloading.
     */
    on?(event: "changed", listener: () => void): Resource;
}

/**
 * Returns a boxed value.
 * In development node, the box value can be changed and observed.
 */
export function createBox<T>(value: T): ObservableBox<T> {
    if (import.meta.env.DEV) {
        return new BoxImpl(value);
    } else {
        return {
            value
        };
    }
}

class BoxImpl<T> extends EventEmitter<{ changed: void }> implements Required<ObservableBox<T>> {
    value: T;

    constructor(value: T) {
        super();
        this.value = value;
    }

    setValue(value: T): void {
        this.value = value;
        this.emit("changed");
    }
}
