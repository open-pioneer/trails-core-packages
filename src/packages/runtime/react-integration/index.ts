// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
export {
    useServiceInternal,
    useServicesInternal,
    usePropertiesInternal,
    useIntlInternal,
    type UseServiceOptions
} from "./hooks";
