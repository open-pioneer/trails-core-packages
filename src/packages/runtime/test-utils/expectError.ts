// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
export function expectError(impl: () => unknown): Error {
    try {
        impl();
        throw new Error("expected error!");
    } catch (e) {
        if (e instanceof Error) {
            return e;
        }
        throw new Error("unexpected error value, not an instance of Error");
    }
}

export function expectAsyncError(impl: () => Promise<unknown>): Promise<Error> {
    const promise = impl();
    return promise.then(
        () => {
            throw new Error("expected error!");
        },
        (e) => {
            if (e instanceof Error) {
                return e;
            }
            throw new Error("unexpected error value, not an instance of Error");
        }
    );
}
