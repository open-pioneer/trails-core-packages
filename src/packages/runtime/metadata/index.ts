// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { ServiceConstructor } from "../Service";
import { ObservableBox } from "./ObservableBox";

/**
 * Application metadata.
 * Typically generated by the build system.
 */
export interface ApplicationMetadata {
    /**
     * Package metadata.
     * Metadata structures contain information about services
     * that are needed during runtime.
     *
     * Services provided by packages in this attribute will be started
     * as necessary and can be referenced during runtime.
     */
    packages?: Record<string, PackageMetadata>;

    /**
     * Styles for UI component.
     */
    styles?: ObservableBox<string>;

    /**
     * Locales supported by the application.
     */
    locales?: string[];

    /**
     * Loads messages for the given locale.
     * @returns A record of packageName -> (messageId, messageTemplate) entries.
     */
    loadMessages?: (locale: string) => Promise<Record<string, Record<string, string>>>;
}

/**
 * Describes a package to the runtime.
 */
export interface PackageMetadata {
    /** The globally unique name of the package. */
    name: string;

    /**
     * Collection of services defined in the package.
     *
     * Key: unique service name within the package.
     * Value: service metadata.
     */
    services?: Record<string, ServiceMetadata>;

    /**
     * Metadata about the UI Components contained in this package.
     */
    ui?: UiMetadata;

    /**
     * Metadata about properties of this package.
     */
    properties?: Record<string, PropertyMetadata>;
}

/**
 * Describes a service to the runtime.
 */
export interface ServiceMetadata {
    /** The unique name of the service in its package. */
    name: string;

    /** Service constructor responsible for creating a new instance. */
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    clazz: ServiceConstructor<any>;

    /**
     * Collection of references to other services.
     * The runtime will inject the required services when possible
     * or provide a diagnostic if not.
     *
     * Key: reference name.
     * Value: reference metadata.
     */
    references?: Record<string, InterfaceReferenceMetadata>;

    /**
     * Lists interfaces provided by this service.
     */
    provides?: ProvidedInterfaceMetadata[];
}

/**
 * Describes a service reference to the runtime.
 */
export interface InterfaceReferenceMetadata {
    /** The name of the referenced interface. */
    name: string;

    /** Additional qualifier to disambiguate interface requirements. */
    qualifier?: string | undefined;

    /** Inject all implementations as an array instead of a specific one. */
    all?: boolean;
}

/**
 * Describes a provided interface to the runtime.
 */
export interface ProvidedInterfaceMetadata {
    /** The name of the provided interface. */
    name: string;

    /** Additional qualifier for disambiguation of multiple interface implementations. */
    qualifier?: string | undefined;
}

/**
 * Describes the UI to the runtime.
 */
export interface UiMetadata {
    /**
     * List of interface names required by the UI.
     * Interfaces in this list can be used by calling the `useService` hook
     * from a react component.
     */
    references?: InterfaceReferenceMetadata[];
}

/**
 * Metadata about a single property in a package.
 */
export interface PropertyMetadata {
    /** Default json value defined by the package. */
    value: unknown;

    /** Whether the property must be specified by the application. */
    required?: boolean;
}

export { type ObservableBox, createBox } from "./ObservableBox";
