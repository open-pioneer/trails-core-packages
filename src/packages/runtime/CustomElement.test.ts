// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0

/**
 * @vitest-environment happy-dom
 */
import { isAbortError } from "@open-pioneer/core";
import {
    defineComponent,
    renderComponent,
    renderComponentShadowDOM
} from "@open-pioneer/test-utils/web-components";
import { findByTestId, waitFor } from "@testing-library/dom";
import { createElement } from "react";
import { expect, it, describe, vi, afterEach } from "vitest";
import { ApiExtension, ApiMethods, ApplicationContext, ApplicationLifecycleListener } from "./api";
import {
    ApplicationElement,
    ApplicationElementConstructor,
    createCustomElement,
    CustomElementOptions
} from "./CustomElement";
import { createBox } from "./metadata";
import { usePropertiesInternal } from "./react-integration";
import { ServiceOptions } from "./Service";
import { expectAsyncError } from "./test-utils/expectError";
import { Portal } from "@chakra-ui/react";

/** Hidden properties available during development / testing */
interface InternalElementType extends ApplicationElement {
    $inspectElementState?(): any;
}

afterEach(() => {
    vi.restoreAllMocks();
});

describe("simple rendering", function () {
    const SIMPLE_STYLE = ".test { color: red }";
    const SIMPLE_ELEM = createCustomElement({
        component: () => createElement("div", { className: "test" }, "hello world"),
        appMetadata: {
            styles: createBox(SIMPLE_STYLE)
        }
    });
    customElements.define("simple-elem", SIMPLE_ELEM);

    it("should return html element", () => {
        expect(new SIMPLE_ELEM()).toBeInstanceOf(HTMLElement);
    });

    it("should render html", async () => {
        const { queries } = await renderComponentShadowDOM("simple-elem");
        const div = await queries.findByText("hello world");
        expect(div.className).toBe("test");
    });

    it("should render user styles", async () => {
        const { shadowRoot } = await renderComponentShadowDOM("simple-elem");

        const styles = await waitFor(() => {
            const styles = shadowRoot.querySelectorAll("style");
            if (styles.length === 0) {
                throw new Error("No styles found");
            }
            return Array.from(styles);
        });

        const trailsStyles = styles.filter((styleNode) => {
            return (
                styleNode.innerHTML.includes("all:initial") ||
                styleNode.innerHTML.includes(".test{")
            );
        });
        expect(trailsStyles).toMatchInlineSnapshot(`
          [
            <style
              data-emotion="css-global"
              data-s=""
            >
              
              @layer base{:host{all:initial;display:block;}}
            </style>,
            <style
              data-emotion="css-global"
              data-s=""
            >
              
              .test{color:red;}
            </style>,
          ]
        `);
    });

    it("should clean up its content when removed from the dom", async () => {
        const { node, shadowRoot, innerContainer } = await renderComponentShadowDOM("simple-elem");
        node.remove();

        // Wait until divs are gone
        await waitFor(() => {
            const div = innerContainer.querySelector("div");
            if (div) {
                throw new Error("content still not destroyed");
            }
        });

        expect(shadowRoot.innerHTML).toBe("");
    });
});

it("uses mode 'open' for the internal shadow root", async () => {
    function TestComponent() {
        return createElement("span", undefined, "Hello World");
    }

    const elem = createCustomElement({
        component: TestComponent
    });
    const { node } = await renderComponent(elem);
    expect(node.shadowRoot).toBeTruthy();
});

it("allows customization of package properties", async () => {
    const elem = createCustomElement({
        component: function TestComponent() {
            const properties = usePropertiesInternal("test");
            return createElement("span", undefined, properties.greeting as string);
        },
        appMetadata: {
            packages: {
                test: {
                    name: "test",
                    properties: {
                        greeting: {
                            value: "Hello World"
                        }
                    }
                }
            }
        },
        config: {
            properties: {
                test: {
                    greeting: "Hello User"
                }
            }
        }
    });

    const { queries } = await renderComponentShadowDOM(elem);
    const span = await queries.findByText("Hello User");
    expect(span.tagName).toBe("SPAN");
});

it("allows customization of package properties through a callback", async () => {
    const elem = createCustomElement({
        component: function TestComponent() {
            const properties = usePropertiesInternal("test");
            return createElement("span", undefined, properties.greeting as string);
        },
        appMetadata: {
            packages: {
                test: {
                    name: "test",
                    properties: {
                        greeting: {
                            value: "Hello World"
                        }
                    }
                }
            }
        },
        config: {
            properties: {
                test: {
                    greeting: "Hello User"
                }
            }
        },
        // properties from this callback take precedence
        async resolveConfig() {
            return {
                properties: {
                    test: {
                        greeting: "Bye User"
                    }
                }
            };
        }
    });

    const { queries } = await renderComponentShadowDOM(elem);
    const span = await queries.findByText("Bye User");
    expect(span.tagName).toBe("SPAN");
});

describe("application context", () => {
    it("reports the shadow dom if enabled", async () => {
        const { element, getNodes } = getNodesHelper(true);

        const {
            node: actualHostElement,
            shadowRoot: actualShadowRoot,
            innerContainer: actualContainer
        } = await renderComponentShadowDOM(element);

        const nodes = getNodes();
        expect(nodes.hostElement).toBe(actualHostElement);
        expect(nodes.shadowRoot).toBe(actualShadowRoot);
        expect(nodes.rootNode).toBe(actualShadowRoot);
        expect(nodes.container).toBe(actualContainer);
    });

    it("reports the document if shadow root is disabled", async () => {
        const { element, getNodes } = getNodesHelper(false);

        const { node: actualHostElement, innerContainer: actualContainer } =
            await renderComponentWithoutShadowRoot(element);

        const nodes = getNodes();
        expect(nodes.hostElement).toBe(actualHostElement);
        expect(nodes.shadowRoot).toBe(undefined);
        expect(nodes.rootNode).toBe(document);
        expect(nodes.container).toBe(actualContainer);
    });

    function getNodesHelper(enableShadowRoot: boolean) {
        let hostElement: HTMLElement | undefined;
        let rootNode: Document | ShadowRoot | undefined;
        let shadowRoot: ShadowRoot | undefined;
        let container: HTMLElement | undefined;

        class TestService {
            constructor(options: ServiceOptions<{ ctx: ApplicationContext }>) {
                const ctx = options.references.ctx;
                hostElement = ctx.getHostElement();
                rootNode = ctx.getRoot();
                shadowRoot = ctx.getShadowRoot();
                container = ctx.getApplicationContainer();
            }
        }

        const element = createCustomElement({
            advanced: {
                enableShadowRoot
            },
            appMetadata: {
                packages: {
                    test: {
                        name: "test",
                        services: {
                            testService: {
                                name: "testService",
                                clazz: TestService,
                                references: {
                                    ctx: {
                                        name: "runtime.ApplicationContext"
                                    }
                                },
                                provides: [
                                    {
                                        name: "runtime.AutoStart"
                                    }
                                ]
                            }
                        }
                    }
                }
            }
        });

        return {
            element,
            getNodes() {
                return {
                    hostElement,
                    rootNode,
                    shadowRoot,
                    container
                };
            }
        };
    }
});

describe("element API", () => {
    it("should throw an error when trying to use the element's API without mounting it first", async () => {
        const elem = createCustomElement({});
        const tag = defineComponent(elem);
        const node = document.createElement(tag) as ApplicationElement;
        const error = await expectAsyncError(() => node.when());
        expect(error).toMatchSnapshot();
    });

    it("should provide an empty API by default", async () => {
        const elem = createCustomElement({});
        const { node } = await renderComponent(elem);
        const api = await (node as ApplicationElement).when();
        expect(api).toEqual({});
    });

    it("throws an error when the component is unmounted while waiting for the API", async function () {
        const elem = createCustomElement({});
        const tag = defineComponent(elem);
        const node = document.createElement(tag) as ApplicationElement;
        document.body.appendChild(node);

        const err = expectAsyncError(() => node.when());
        node.remove();

        const error = await err;
        expect(isAbortError(error)).toBe(true);
    });

    it("should allow services to provide an API", async () => {
        const events: string[] = [];
        class Extension implements ApiExtension {
            async getApiMethods(): Promise<ApiMethods> {
                return {
                    add(x: number, y: number) {
                        events.push("add");
                        return x + y;
                    },
                    otherMethod() {
                        events.push("otherMethod");
                    }
                };
            }
        }

        const elem = createCustomElement({
            appMetadata: {
                packages: {
                    test: {
                        name: "test",
                        services: {
                            testService: {
                                name: "testService",
                                provides: [
                                    {
                                        name: "integration.ApiExtension"
                                    }
                                ],
                                clazz: Extension
                            }
                        }
                    }
                }
            }
        });

        const { node } = await renderComponent(elem);
        const api = await (node as ApplicationElement).when();
        expect(Object.keys(api).sort()).toEqual(["add", "otherMethod"]);
        expect(events).toEqual([]);

        const sum = api.add!(3, 4);
        expect(sum).toEqual(7);
        expect(events).toEqual(["add"]);

        api.otherMethod!();
        expect(events).toEqual(["add", "otherMethod"]);
    });
});

describe("application lifecycle events", function () {
    it("signals 'after-start' and 'before-stop' events", async function () {
        const events: string[] = [];
        class Listener implements ApplicationLifecycleListener {
            afterApplicationStart() {
                events.push("start");
            }

            beforeApplicationStop() {
                events.push("stop");
            }
        }

        const elem = createCustomElement({
            appMetadata: {
                packages: {
                    test: {
                        name: "test",
                        services: {
                            Listener: {
                                name: "Listener",
                                provides: [
                                    {
                                        name: "runtime.ApplicationLifecycleListener"
                                    }
                                ],
                                clazz: Listener
                            }
                        }
                    }
                }
            }
        });

        const { node, innerContainer } = await renderComponentShadowDOM(elem);
        expect(events).toEqual(["start"]);

        // Wait until divs are gone
        node.remove();
        await waitFor(() => {
            const div = innerContainer.querySelector("div");
            if (div) {
                throw new Error("content still not destroyed");
            }
        });

        expect(events).toEqual(["start", "stop"]);
    });

    it("does not signal 'before-stop' when start fails", async function () {
        const errorSpy = vi.spyOn(console, "error").mockImplementation(() => undefined);

        const events: string[] = [];
        class Listener implements ApplicationLifecycleListener {
            afterApplicationStart() {
                events.push("start");
            }

            beforeApplicationStop() {
                events.push("stop");
            }
        }

        const elem = createCustomElement({
            async resolveConfig() {
                throw new Error("help!");
            },
            appMetadata: {
                packages: {
                    test: {
                        name: "test",
                        services: {
                            Listener: {
                                name: "Listener",
                                provides: [
                                    {
                                        name: "runtime.ApplicationLifecycleListener"
                                    }
                                ],
                                clazz: Listener
                            }
                        }
                    }
                }
            }
        });

        // App starts up and immediately stops because of the error in `resolveConfig` above.
        const { node } = await renderComponent(elem);
        await waitFor(() => {
            const state = (node as InternalElementType).$inspectElementState?.().state;
            if (state !== "error") {
                throw new Error(`App did not reach destroyed state.`);
            }
        });

        expect(events).toEqual([]);
        expect(errorSpy).toMatchInlineSnapshot(`
          [MockFunction error] {
            "calls": [
              [
                "#1",
                [Error: help!],
              ],
              [
                "#2",
                [Error: runtime:config-resolution-failed: Failed to resolve application properties.],
              ],
            ],
            "results": [
              {
                "type": "return",
                "value": undefined,
              },
              {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        `);
    });
});

describe("service autostart", function () {
    it("starts all services implementing the AutoStart interface", async function () {
        const events: string[] = [];
        class ListenerA {
            constructor() {
                events.push("A");
            }
        }
        class ListenerB {
            constructor() {
                events.push("B");
            }
        }

        const elem = createCustomElement({
            appMetadata: {
                packages: {
                    test: {
                        name: "test",
                        services: {
                            ListenerA: {
                                name: "ListenerA",
                                provides: [
                                    {
                                        name: "runtime.AutoStart"
                                    }
                                ],
                                clazz: ListenerA
                            },
                            ListenerB: {
                                name: "ListenerB",
                                provides: [
                                    {
                                        name: "runtime.AutoStart"
                                    }
                                ],
                                clazz: ListenerB
                            }
                        }
                    }
                }
            }
        });

        await renderComponentShadowDOM(elem);
        events.sort();
        expect(events).toEqual(["A", "B"]);
    });
});

describe("i18n support", function () {
    afterEach(() => {
        vi.restoreAllMocks();
    });

    it("chooses locale from the supported browser locales (german)", async () => {
        const spy = vi.spyOn(window.navigator, "languages", "get");
        spy.mockReturnValue(["de-DE", "de", "en"]);

        const api = await launchApp();
        const { locale, message } = await api.getLocaleInfo();
        expect(locale).toBe("de-DE");
        expect(message).toBe("Hallo Welt");
    });

    it("chooses locale from the supported browser locales (english)", async () => {
        const spy = vi.spyOn(window.navigator, "languages", "get");
        spy.mockReturnValue(["en-US", "en"]);

        const api = await launchApp();
        const { locale, message } = await api.getLocaleInfo();
        expect(locale).toBe("en-US");
        expect(message).toBe("Hello world");
    });

    it("supports forcing to a specific locale", async () => {
        const spy = vi.spyOn(window.navigator, "languages", "get");
        spy.mockReturnValue(["en-US", "en"]);

        const api = await launchApp({
            config: {
                locale: "de-simple"
            }
        });
        const { locale, message } = await api.getLocaleInfo();
        expect(locale).toBe("de-simple");
        expect(message).toBe("Hallo Welt (einfach)");
    });

    it("supports forcing to a specific locale with resolveConfig()", async () => {
        const spy = vi.spyOn(window.navigator, "languages", "get");
        spy.mockReturnValue(["en-US", "en"]);

        const api = await launchApp({
            config: {
                locale: "en" // overwritten by resolveConfig
            },
            resolveConfig() {
                return Promise.resolve({
                    locale: "de-simple"
                });
            }
        });
        const { locale, message } = await api.getLocaleInfo();
        expect(locale).toBe("de-simple");
        expect(message).toBe("Hallo Welt (einfach)");
    });

    it("supports restarting with a different locale", async () => {
        // Hide i18n warnings
        vi.spyOn(console, "warn").mockImplementation(() => undefined);

        const spy = vi.spyOn(window.navigator, "languages", "get");
        spy.mockReturnValue(["en-US", "en"]);

        const api = await launchApp({
            config: {
                locale: "en-US"
            }
        });
        const { locale, message, supportedLocales } = await api.getLocaleInfo();
        expect(locale).toBe("en-US");
        expect(message).toBe("Hello world");
        expect(supportedLocales).toMatchInlineSnapshot(`
          [
            "de",
            "en",
            "de-simple",
          ]
        `);

        // Changes locale and has priority over the `config` above
        await api.setLocale("de-DE");
        await waitFor(async () => {
            const { locale } = await api.getLocaleInfo();
            if (locale !== "de-DE") {
                throw new Error("locale not changed yet");
            }
        });
        const { locale: newLocale, message: newMessage } = await api.getLocaleInfo();
        expect(newLocale).toBe("de-DE");
        expect(newMessage).toBe("Hallo Welt");

        // Check that invalid locale throws an error
        await expect(() => api.setLocale("zh-CN")).rejects.toThrowErrorMatchingInlineSnapshot(
            `[Error: runtime:unsupported-locale: Unsupported locale 'zh-CN' (supported locales: de, en, de-simple).]`
        );
    });

    interface I18nAppApi {
        getLocaleInfo(): Promise<{ locale: string; message: string; supportedLocales: string[] }>;
        setLocale(newLocale: string): Promise<void>;
    }

    /**
     * Runs an app with mocked services and i18n and returns the inner locale + translated message.
     */
    async function launchApp(options?: Partial<CustomElementOptions>): Promise<I18nAppApi> {
        class TestService implements ApiExtension {
            private ctx: ApplicationContext;
            private locale: string;
            private message: string;

            constructor(options: ServiceOptions<{ ctx: ApplicationContext }>) {
                const ctx = options.references.ctx;
                this.ctx = ctx;
                this.locale = ctx.getLocale();
                this.message = options.intl.formatMessage({ id: "greeting" });
            }

            async getApiMethods(): Promise<ApiMethods> {
                return {
                    getLocaleInfo: () => {
                        return {
                            locale: this.locale,
                            message: this.message,
                            supportedLocales: Array.from(this.ctx.getSupportedLocales())
                        };
                    },
                    setLocale: (newLocale: string) => {
                        this.ctx.setLocale(newLocale);
                    }
                };
            }
        }

        const elem = createCustomElement({
            ...options,
            appMetadata: {
                packages: {
                    test: {
                        name: "test",
                        services: {
                            testService: {
                                name: "testService",
                                clazz: TestService,
                                references: {
                                    ctx: {
                                        name: "runtime.ApplicationContext"
                                    }
                                },
                                provides: [
                                    {
                                        name: "integration.ApiExtension"
                                    }
                                ]
                            }
                        }
                    }
                },
                locales: ["de", "en", "de-simple"],
                async loadMessages(locale) {
                    switch (locale) {
                        case "en":
                            return {
                                test: {
                                    greeting: "Hello world"
                                }
                            };
                        case "de":
                            return {
                                test: {
                                    greeting: "Hallo Welt"
                                }
                            };
                        case "de-simple":
                            return {
                                test: {
                                    greeting: "Hallo Welt (einfach)"
                                }
                            };
                    }
                    throw new Error("Unsupported locale: " + locale);
                }
            }
        });

        const { node } = await renderComponentShadowDOM(elem);
        const result = {
            async getLocaleInfo() {
                const api = await (node as ApplicationElement).when();
                return api.getLocaleInfo!();
            },

            async setLocale(newLocale: string) {
                const api = await (node as ApplicationElement).when();
                api.setLocale!(newLocale);
            }
        };
        return result;
    }
});

it("renders an error screen when the app fails to start", async () => {
    vi.spyOn(console, "error").mockImplementation(() => undefined);

    const elem = createCustomElement({
        async resolveConfig() {
            throw new Error("help!");
        }
    });

    const { node, shadowRoot } = await renderComponentShadowDOM(elem);
    await waitFor(() => {
        const state = (node as InternalElementType).$inspectElementState?.().state;
        if (state !== "error") {
            throw new Error(`App did not reach error state.`);
        }
    });

    const errorScreen = shadowRoot.querySelector("div.pioneer-root")!;
    expect(errorScreen).toBeTruthy();

    const classes = Array.from(errorScreen.classList);
    expect(classes).toContain("pioneer-root-error-screen");

    await waitFor(() => {
        const includesErrorText = Array.from(errorScreen?.children ?? []).some((child) =>
            child.textContent?.includes("Error")
        );
        expect(includesErrorText).toBe(true);
    });
});

it("can render without a shadow root", async () => {
    function AppUI() {
        const portalContent = createElement(
            Portal,
            undefined,
            createElement(
                "div",
                { className: "portal-content", "data-testid": "portal-content" },
                "Hello from Portal"
            )
        );

        return createElement(
            "div",
            {
                "data-testid": "main-content"
            },
            "Hello World",
            portalContent
        );
    }

    const elem = createCustomElement({
        component: AppUI,
        advanced: {
            enableShadowRoot: false
        }
    });
    const { innerContainer } = await renderComponentWithoutShadowRoot(elem);

    const mainContent = await findByTestId(innerContainer, "main-content");
    expect(mainContent.textContent).toBe("Hello World");
    expect(mainContent.parentElement).toBe(innerContainer);

    // Important: portal content must be a child of the `.pioneer-root`.
    // This is currently implemented using a patch against ark ui.
    const portalContent = await findByTestId(innerContainer, "portal-content");
    expect(portalContent.textContent).toBe("Hello from Portal");
    expect(portalContent.parentElement).toBe(innerContainer);

    const styles = Array.from(document.head.querySelectorAll("style"));
    const rule = styles.find((s) => s.innerHTML.includes(".pioneer-root"));
    expect(rule).toBeTruthy(); // Styles are mounted into the document head
});

async function renderComponentWithoutShadowRoot(element: ApplicationElementConstructor) {
    const { node } = await renderComponent(element);
    const innerContainer = await waitFor(() => {
        const pioneerRoot = node.querySelector(`.pioneer-root`);
        if (!pioneerRoot) {
            throw new Error("Pioneer root node did not mount.");
        }
        return pioneerRoot as HTMLElement;
    });

    return {
        node,
        innerContainer
    };
}
