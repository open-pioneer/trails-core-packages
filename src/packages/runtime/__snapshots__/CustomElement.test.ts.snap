// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`element API > should throw an error when trying to use the element's API without mounting it first 1`] = `[Error: runtime:element-not-mounted: Cannot use the application's API because the HTML element has not yet been mounted into the DOM.]`;

exports[`simple rendering > should render user styles 1`] = `
[
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    @layer base{:host{all:initial;display:block;}}
  </style>,
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    .test{color:red;}
  </style>,
]
`;
